src/explore-education-statistics-common/src/modules/table-tool/components/FilterHierarchyOptions.tsx

```diff
@@ -1,78 +1,155 @@
-  const childOptionIds = filterHierarchy[level]?.hierarchy[optionId] ?? [];
+  const [searchTerm, setSearchTerm] = useState<string>('');
+
+  const filteredOptions = useMemo(() => {
+    if (!optionTree.options) return [];
+    if (!searchTerm) return optionTree.options;
+    return optionTree.options.filter(option =>
+      option.label.toLowerCase().includes(searchTerm.trim().toLowerCase()),
+    );
+  }, [optionTree, searchTerm]);
+
+  const { childOptionIds, hasAllSelected } = useMemo(() => {
+    const childIds = filteredOptions?.map(({ value }) => value) ?? [];
+    return {
+      childOptionIds: childIds,
+      hasAllSelected: childIds.every(childOptionId =>
+        selectedValues.includes(childOptionId),
+      ),
+    };
+  }, [selectedValues, filteredOptions]);
+
+  const isExpanded = expandedOptionsList.includes(optionTree.value);
+  const { setValue } = useFormContext();
+
+  const toggleSelectAll = useCallback(() => {
+    if (hasAllSelected) {
+      return setValue(
+        name,
+        selectedValues.filter(
+          selectedValue => !childOptionIds.includes(selectedValue),
+        ),
+      );
+    }
+    return setValue(
+      name,
+      Array.from(new Set([...selectedValues, ...childOptionIds])),
+    );
+  }, [childOptionIds, setValue, hasAllSelected, selectedValues, name]);
 
   return (
     <div
-      className={classNames('govuk-checkboxes', {
-        'govuk-checkboxes--small': true,
-      })}
+      className={classNames(
+        'govuk-checkboxes',
+        'govuk-checkboxes--small',
+        styles.option,
+        {
+          [styles.option__expanded]: isExpanded,
+        },
+      )}
     >
-      <div className="govuk-!-margin-left-8 govuk-!-margin-bottom-2">
-        {hideLowerTotals &&
-        level !== 0 &&
-        optionLabel.toLocaleLowerCase() === 'total' &&
-        hideLowerTotals ? null : (
-          <FormField
-            label={optionLabel}
-            id={`${name}-${optionId}`}
-            // @ts-expect-error  `value` required to make checkboxes unique
-            value={optionId}
-            name={name}
-            as={FormCheckbox}
-            checked={selectedValues.includes(optionId)}
-            disabled={disabled}
-          />
-        )}
-        {childOptionIds.map(childOptionId => (
-          <FilterHierarchyOptions
-            optionId={childOptionId}
-            key={childOptionId}
-            optionLabelsMap={optionLabelsMap}
-            filterHierarchy={filterHierarchy}
-            name={name}
-            disabled={disabled}
-            tiersTotal={tiersTotal}
-            selectedValues={selectedValues}
-            level={level + 1}
-          />
-        ))}
+      <div className={styles.checkbox}>
+        <FormField
+          hint={optionTree.filterLabel}
+          label={optionTree.label}
+          id={`${name}-${optionTree.value}`}
+          // @ts-expect-error  `value` required to make checkboxes unique
+          value={optionTree.value}
```

Comment:
I really think there's got to be a better way to do this. It's like react hook form isn't assigning the value as it does for other checkboxes so there's got to be something causing that. But, that said, I've had a quick go and can't figure it out.

Reply:
There definitely was, it took a while for me to even get this working, I've now since improved it I'd say, this new implementation seems so obvious, not sure how I missed it.